text/0000-jsbox.md

```diff
@@ -0,0 +1,341 @@
+* Multiple `JsBox`, even if they are unrelated, could not be borrowed at the same time because there is no way to statically prove they do not point to the same underlying data
+* Impairs ergonomics on APIs that do not require mutable borrowing since `Deref` cannot be implemented
+
+### `JsCell`
+
+Neon could provide a `JsCell` type that is identical to `JsBox`, but also wraps the underlying `T` in a `std::cell::RefCell`. This allows the user to also mutably borrow the value.
+
+```rust
+impl<T> JsCell where T: Send + 'static {
+    /// This should only fail if the VM is in a throwing state or is
+    /// in the process of stopping.
+    pub fn new<'a, C: Context<'a>>(
+        cx: &mut C,
+        v: T,
+    ) -> NeonResult<JsCell>;
+    /// May panic if already mutably borrowed
+    pub fn borrow<'a, C: Context<'a>>(
+        &self,
+        cx: &mut C,
+    ) -> Ref<'a, T>;
+    pub fn try_borrow<'a, C: Context<'a>>(
+        &self,
+        cx: &mut C,
+    ) -> Result<Ref<'a, T>, BorrowError>;
+    /// May panic if already borrowed
+    pub fn borrow_mut<'a, C: Context<'a>>(
+        &self,
+        cx: &mut C,
+    ) -> RefMut<'a, T>;
+    pub fn try_borrow_mut<'a, C: Context<'a>>(
+        &self,
+        cx: &mut C,
+    ) -> Result<RefMut<'a, T>, BorrowMutError>;
+    /// The most common `RefCell` methods are provided on `JsCell` as a
+    /// convenience. Other methods may can be accessed by calling `as_cell`
+    pub fn as_cell<'a, C: Context<'a>>(
+        &self,
+        cx: &mut C,
+    ) -> &RefCell<T>;
+}
+```
+
+This approach has several significant drawbacks:
+
+* Increases the API substantial
+* Adds an abstraction that not all users might need
+* Prevents implementing `Deref`
+
+Instead, Neon can lean on documentation to guide users towards `RefCell` if they need interior mutability.
+
+### Naming
+
+Several other names were considered:
+
+* `JsExternal`. Matches the name in N-API documentation (external reference); however, `cx.external(v)` might be confusing.
+* `JsRef`. Also, matches N-API but, does not match Rust naming conventions.
+* `JsCell`. Only appropriate for the `RefCell` approach.
+
+Overall, `JsBox<T>` behaves very similar to a `std::boxed::Box<T>`.
+
+### Leverage existing borrow API
+
+An existing [`neon::borrow`](https://docs.rs/neon/0.4.0/neon/borrow/index.html) implementation exists. However, the current implementation has multiple soundness holes:
+
+* `Buffer` can be mutably borrowed multiple times
+* `Lock` does not prevent overlapping memory regions
+* Multiple [`Lock`](https://docs.rs/neon/0.4.0/neon/context/struct.Lock.html) can be created allowing aliased mutable borrows
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Should we implement this for the legacy backend? We likely will not since this is a brand new API.
+- Is it acceptable to let `JsBox::new` panic since it only fails when throwing or shutting down?
```

Comment:
Yes, these are both highly rare cases, which is the spirit of panic.

Reply:
Currently the RFC defines the APIs as returning a `JsResult`, but that's inconsistent with other methods. I think that should be changed to return an unwrapped `JsBox<_>`.

If we ever decide in the future that we want this potential failure exposed directly, it would be best done all at once for consitency.

Reply:
I feel comfortable making all the APIs return an unwrapped `JsBox<_>`. `NeonResult` is meant to represent whether the operation caused a JS exception. Failing because it was already in a throwing state is basically a programmer error that we happen not to be able to enforce through the type system and therefore a panic; the VM shutting down is extremely rare and unrelated to triggering a JS exception.